Strategy prep

mzolkiewski opened the pull request.

wodsuz commented on src/controllers/ScheduleController.ts:
```diff
@@ -13,10 +14,7 @@ export class ScheduleController {
-      );
+      const stored = await this.scheduleService.schedule(req.body, normalizeScheduleRequestParams(req.query));
       res.json({
         id: stored._id,
```
here id doesnt have a type, maybe we can define a type for it?

mzolkiewski replied:
It kinda does, since it's the same type as `stored._id` and that one is actually a mongodb id (i.e. string). but it's true that we could define some return types for the whole response. however, express does not let you do type arguments for json response (`res.json<type>({ ... })`) so it's let's leave it like that for now

wodsuz commented on src/models/StrategyPrepSchema.ts:
```diff
@@ -0,0 +1,84 @@
+import { createLogger } from '../logger';
+import { IStrategyPrep } from './Models';
+
+const logger = createLogger('StrategyPrepSchema');
```
i didnt see the usage of this logger on this file. maybe a logger.debug('StrategyPrepSchema') is good?

mzolkiewski replied:
it's just boilerplate for when it's needed. all other models have it. it's to remind myself that if I want to log something, I need to use this logger

wodsuz commented on src/services/schedule.ts:
```diff
@@ -37,8 +41,9 @@ export interface IScheduleService {
-    if (params?.source === 'proxy') {
+  public async schedule(request: IScheduleRequest, params: IScheduleParams) {
+    const isProxyRequest = params.source === 'proxy';
+    if (isProxyRequest) {
```
Using this instead of nested if's is a better approach? 

if (isProxyRequest && txCache.get(request.signedTransaction)) {
        throw new Error('Duplicate request');
    }
 if (isProxyRequest) {
      txCache.put(request.signedTransaction, true);
    }

mzolkiewski replied:
You are absolutely right, but actually the whole `schedule` function needs to be refactored to handle the various cases more clearly. I did not have a good idea how to split up the code yet, so I'm leaving the 'status quo' for now

wodsuz commented on src/services/schedule.ts:
```diff
@@ -104,41 +84,48 @@ export class ScheduleService implements IScheduleService {
     // otherwise we have a problem
     if (!transactionExists) {
-      let duplicate = await this.findBySignedTransaction(request.signedTransaction);
+      const duplicate = await findBySignedTransaction(request.signedTransaction);
```
Another nested if. But this time there's a varaible in between making it difficult to simplify. My recommendation is, putting the variable outside might cause lower performance and time consumption tho,

```
const duplicate = await findBySignedTransaction(request.signedTransaction);
if (!transactionExists && duplicate) {
        throw new Error(`Duplicate transaction ${request.signedTransaction}`);
    }
```

mzolkiewski replied:
Correct, this is ugly. The better solution is to extract the inner logic into a separate function, which I did

wodsuz commented on src/services/schedule.ts:
```diff
@@ -227,29 +214,207 @@ export class ScheduleService implements IScheduleService {
+    }
+    case AssetType.Polkadot:
+    default: {
+      throw new Error('Implementme!');
```
tiny typo Implement!

mzolkiewski replied:
It's actually supposed to be like this. It says 'implement me!', i.e. the case is asking to be implemented. It's just a convention

wodsuz reviewed (approved):
I have found few nested ifs and thought quite a lot about best practice but I am still not sure tbh.

mzolkiewski commented on src/models/StrategyPrepSchema.ts:
```diff
@@ -0,0 +1,84 @@
+  timeCondition: {
+    type: Number,
+  },
+  timeConditionTZ: {
```
yes, this mirrors the property in `Scheduled.ts` model

mzolkiewski commented:
> I have found few nested ifs and thought quite a lot about best practice but I am still not sure tbh.

Thanks, I addressed some of your comments - I've already done quite a lot of refactoring to improve the code but there's still a bunch to be done
